File: pipeline/generate.py
"""
Orchestrator — run with:  python -m pipeline.generate

Flow:
  retrieve candidates  ->  select top stories (1 LLM call)
  ->  scrape each story's sources  ->  analyse each story (1 LLM call each)
  ->  write data/YYYY-MM-DD.json and refresh data/index.json

Total LLM calls per run ≈ TOP_N + 1. At TOP_N=8 that's ~9 calls/day, far below
the free-tier daily allowance.
"""

from __future__ import annotations
import datetime as dt
import json
import pathlib

from . import retrieve, scrape
from .select import select_top_stories
from .analyze import analyze_story

DATA_DIR = pathlib.Path(__file__).resolve().parent.parent / "data"


def _slug(text: str) -> str:
    keep = [c.lower() if c.isalnum() else "-" for c in text]
    s = "".join(keep)
    while "--" in s:
        s = s.replace("--", "-")
    return s.strip("-")[:60] or "story"


def run() -> None:
    today = dt.date.today().isoformat()
    print(f"== Generating deep dives for {today} ==")

    print("1) Retrieving candidates...")
    candidates = retrieve.fetch_candidates()
    if not candidates:
        print("No candidates found — aborting (check feeds/network).")
        return

    print("2) Selecting top stories...")
    selected = select_top_stories(candidates)
    print(f"   selected {len(selected)} stories")

    stories_out = []
    for n, sel in enumerate(selected, 1):
        print(f"3.{n}) Story: {sel.title}")
        articles = []
        for idx in sel.article_indices:
            c = candidates[idx]
            print(f"      scraping {c['source']}...")
            text = scrape.fetch_text(c["url"])
            articles.append({**c, "text": text})

        deep = analyze_story(sel.title, articles)
        if deep is None:
            continue

        sources_used = [
            {"outlet": a["source"], "lean": a["lean"], "region": a["region"],
             "title": a["title"], "url": a["url"]}
            for a in articles if a.get("text")
        ]
        story = deep.model_dump()
        story["id"] = _slug(deep.headline)
        story["region"] = sel.region
        story["sources"] = sources_used
        stories_out.append(story)

    if not stories_out:
        print("No stories produced — aborting write.")
        return

    payload = {
        "date": today,
        "generated_at": dt.datetime.now(dt.timezone.utc).isoformat(),
        "stories": stories_out,
    }

    DATA_DIR.mkdir(exist_ok=True)
    out_path = DATA_DIR / f"{today}.json"
    out_path.write_text(json.dumps(payload, indent=2, ensure_ascii=False))
    print(f"   wrote {out_path}")

    # Refresh the index of available dates (newest first).
    dates = sorted(
        (p.stem for p in DATA_DIR.glob("*.json") if p.stem != "index"),
        reverse=True,
    )
    (DATA_DIR / "index.json").write_text(json.dumps(dates, indent=2))
    print(f"   index updated: {len(dates)} day(s)")
    print("== Done ==")


if __name__ == "__main__":
    run()

File: pipeline/generate.js
import { existsSync, writeFileSync, mkdirSync, readdirSync, readFileSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import { fetchCandidates } from "./retrieve.js";
import { selectTopStories } from "./select.js";
import { fetchText } from "./scrape.js";
import { analyzeStory } from "./analyze.js";
import { structured } from "./llm.js";
import { RECAP_SCHEMA } from "./schema.js";
import { TOP_N } from "./sources.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const DATA_DIR = resolve(__dirname, "../data");

function slug(text) {
  let s = text.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/-+/g, "-").replace(/^-|-$/g, "");
  return s.slice(0, 60) || "story";
}

function refreshIndex() {
  const entries = readdirSync(DATA_DIR)
    .filter(f => /^\d{4}-\d{2}-\d{2}\.json$/.test(f))
    .map(f => {
      const date = f.replace(".json", "");
      try {
        const { stories } = JSON.parse(readFileSync(`${DATA_DIR}/${f}`, "utf8"));
        return { date, count: stories?.length ?? 0 };
      } catch {
        return { date, count: 0 };
      }
    })
    .sort((a, b) => b.date.localeCompare(a.date));
  writeFileSync(`${DATA_DIR}/index.json`, JSON.stringify(entries, null, 2));
  console.log(`   index updated: ${entries.length} day(s)`);
}

async function run() {
  const today = new Date().toISOString().slice(0, 10);
  const outPath = `${DATA_DIR}/${today}.json`;

  if (existsSync(outPath) && !process.env.FORCE_REGEN) {
    console.log(`== data/${today}.json already exists — nothing to do.`);
    console.log(`   (set FORCE_REGEN=1 to regenerate today's briefing)`);
    refreshIndex();
    return;
  }

  console.log(`== Generating deep dives for ${today} (target: ${TOP_N} stories) ==`);

  console.log("1) Retrieving candidates...");
  const candidates = await fetchCandidates();
  if (!candidates.length) {
    console.log("No candidates found — aborting (check feeds/network).");
    return;
  }

  console.log("2) Selecting stories (primaries + reserves)...");
  const selected = await selectTopStories(candidates);
  console.log(`   pool: ${selected.length} stories (${TOP_N} target + up to ${selected.length - TOP_N} reserves)`);

  // ── Backfill loop ────────────────────────────────────────────────────────
  // Work through the pool in priority order. Stop as soon as TOP_N stories
  // have been successfully scraped + analysed. Any reserves beyond that are
  // never even fetched, keeping API usage minimal.
  const storiesOut = [];

  for (let n = 0; n < selected.length; n++) {
    if (storiesOut.length >= TOP_N) break; // ✓ hit the target — stop

    const sel = selected[n];
    const isReserve = n >= TOP_N;
    const label = isReserve ? `reserve-${n - TOP_N + 1}` : String(n + 1);
    console.log(`\n${label}) ${sel.title}`);

    // Scrape all articles for this story
    const articles = [];
    for (const idx of sel.article_indices) {
      const c = candidates[idx];
      console.log(`      scraping ${c.source}...`);
      const text = await fetchText(c.url);
      articles.push({ ...c, text });
    }

    // Skip if nothing scraped (no text to analyse)
    const usable = articles.filter(a => a.text);
    if (!usable.length) {
      console.log(`    ! no usable text — ${isReserve ? "reserve" : "trying next reserve"}`);
      continue;
    }

    let deep;
    try {
      deep = await analyzeStory(sel.title, articles);
    } catch (err) {
      console.log(`    ! analysis failed: ${err.message?.slice(0, 120)} — ${isReserve ? "reserve" : "trying next reserve"}`);
      continue;
    }
    if (!deep) continue;

    const sourcesUsed = articles
      .filter(a => a.text)
      .map(a => ({ outlet: a.source, lean: a.lean, region: a.region, title: a.title, url: a.url }));

    storiesOut.push({
      ...deep,
      id: slug(deep.headline),
      region: sel.region,
      sources: sourcesUsed,
    });

    console.log(`    ✓ story ${storiesOut.length}/${TOP_N} done (${sourcesUsed.length} source${sourcesUsed.length > 1 ? "s" : ""})`);
  }

  if (!storiesOut.length) {
    console.log("No stories produced — aborting write.");
    return;
  }

  if (storiesOut.length < TOP_N) {
    console.log(`\n⚠  Only ${storiesOut.length}/${TOP_N} stories produced — pool exhausted.`);
  }

  const payload = {
    date: today,
    generated_at: new Date().toISOString(),
    stories: storiesOut,
  };

  mkdirSync(DATA_DIR, { recursive: true });
  writeFileSync(outPath, JSON.stringify(payload, null, 2));
  console.log(`\n   wrote ${outPath} (${storiesOut.length} stories)`);

  refreshIndex();
  console.log("== Done ==");
}

run().catch(err => { console.error(err); process.exit(1); });
